tests/testthat/test-slice.R

```diff
@@ -272,6 +272,18 @@ test_that("vec_slice() unclasses input before calling `vec_restore()`", {
 })
 
 test_that("can call `vec_slice()` from `[` methods with shaped objects without infloop", {
+  skip("Until infloop is talked about")
+  # TODO: I don't think it is our job to prevent this.
+  # You can already make this infloop today with non-shaped objects, i.e
+  # `[.vctrs_foobar` = function(x, i, ...) vec_slice(x, i)
+  # x <- structure(1:4, class = "vctrs_foobar")
+  # x[1]
+  # So we shouldn't be special casing shaped objects.
+  # I believe that if your `[` method is going to call `vec_slice()` (i.e. you
+  # are requesting native vctrs slicing), then you need to declare a
+  # `vec_proxy()` method as well to tell vctrs what it needs to be natively
+  # slicing.
+
   local_methods(
     `[.vctrs_foobar` = function(x, i, ...) vec_slice(x, i)
```

Comment:
Ok, so lets talk about this. This infloop check is for:
- A class that has clearly opted into using vctrs, because they call `vec_slice()`
- But isn't using `new_vctr()`, which has a `vec_proxy()` method defined
- And isn't defining their own `vec_proxy()` method

In that case, `vec_slice()` tries to fall back to the `[` method, which causes the infloop.

I don't think that avoiding this infloop is our responsibility. The way we tried to do this was by calling `unclass()` in that original `vec_slice_fallback()` function, but as I showed earlier that defeats the whole purpose of the function.

It it also 100% possible to do this already with non-shaped vectors. i.e. this is with CRAN vctrs:

```r
library(vctrs)
`[.vctrs_foobar` <- function(x, i, ...) vec_slice(x, i)
x <- structure(1:4, class = "vctrs_foobar")
# infloop
x[1]
```

I think that if you are telling vctrs that you want `[` to forward to `vec_slice()`, then you _have_ to tell vctrs what it should be natively slicing by providing a `vec_proxy()` method (if you aren't extending vctrs_vctr).